lightning-block-sync/src/lib.rs

```diff
@@ -0,0 +1,910 @@
+/// not responding to requests for old headers on that fork. However, if one block source is
+/// unreachable this may result in our memory usage growing in accordance with the chain.
+pub struct MicroSPVClient<'a, P, B, CL>
+where P: Poll<'a, B>,
+	  B: DerefMut<Target=dyn BlockSource + 'a> + Sized + Sync + Send,
+	  CL: ChainListener + Sized {
+	chain_tip: (BlockHash, BlockHeaderData),
+	chain_poller: P,
+	backup_block_sources: Vec<B>,
+	blocks_past_common_tip: Vec<BlockHeaderData>,
+	chain_notifier: CL,
+	mainnet: bool
+}
+
+impl<'a, P, B, CL> MicroSPVClient<'a, P, B, CL>
+where P: Poll<'a, B>,
+	  B: DerefMut<Target=dyn BlockSource + 'a> + Sized + Sync + Send,
+	  CL: ChainListener + Sized {
+
+	/// Creates a new `MicroSPVClient` with a chain poller for polling one or more block sources and
+	/// a chain listener for receiving updates of the new chain tip.
+	///
+	/// At least one of the polled `BlockSource`s must provide the necessary headers to disconnect
+	/// from the given `chain_tip` back to its common ancestor with the best chain assuming that its
+	/// height, hash, and chainwork are correct.
+	///
+	/// `backup_block_sources` are never queried unless we learned, via some `block_sources` source
+	/// that there exists a better, valid header chain but we failed to fetch the blocks. This is
+	/// useful when you have a block source which is more censorship-resistant than others but
+	/// which only provides headers. In this case, we can use such source(s) to learn of a censorship
+	/// attack without giving up privacy by querying a privacy-losing block sources.
+	pub fn init(chain_tip: BlockHeaderData, chain_poller: P, backup_block_sources: Vec<B>, chain_notifier: CL, mainnet: bool) -> Self {
+		let blocks_past_common_tip = Vec::new();
+		Self {
+			chain_tip: (chain_tip.header.block_hash(), chain_tip),
+			chain_poller, backup_block_sources, blocks_past_common_tip, chain_notifier, mainnet
+		}
+	}
+
+	/// Check each source for a new best tip and update the chain listener accordingly.
+	/// Returns true if some blocks were [dis]connected, false otherwise.
+	pub async fn poll_best_tip(&mut self) -> bool {
+		macro_rules! sync_chain_monitor {
+			($new_hash: expr, $new_header: expr, $source: expr) => { {
+				let mut blocks_connected = false;
+				match sync_chain_monitor($new_header, &self.chain_tip.1, $source, &mut self.chain_notifier, &mut self.blocks_past_common_tip, self.mainnet).await {
+					Err((_, latest_tip)) => {
+						if let Some(latest_tip) = latest_tip {
+							let latest_tip_hash = latest_tip.header.block_hash();
+							if latest_tip_hash != self.chain_tip.0 {
+								self.chain_tip = (latest_tip_hash, latest_tip);
+								blocks_connected = true;
+							}
+						}
+					},
+					Ok(_) => {
+						self.chain_tip = ($new_hash, $new_header);
+						blocks_connected = true;
+					},
+				}
+				blocks_connected
+			} }
+		}
+
+		match self.chain_poller.poll_chain_tip(self.chain_tip.1).await {
```

Comment:
I'm a little confused by the distinction between chain_poller and the block sources - given many block sources almost certainly will be a part of what we expect to poll for the current best chain, it seems like they shouldn't be separated.

Still, I wonder if we can't move a bunch of the source logic into the Poll trait - instead of multiplexing between sources here, make the Poll trait (maybe renamed to Multiplexer or something) implement `BlockSource` given a list of primary and backup `BlockSource`s. This may also be an opportunity to simplify the initial PR, since, assuming its doable, we could implement MicroSPVClient against a single known block source and not have to do a lot of the tracking we do here, just implementing the chain walking.

Reply:
Right, so we chatted about this a bit offline, but I'm still a bit unsure about the separation here. I went back and reminded myself of how it was done in 614, but I think maybe we want to expand, or at least have more flexibility, around disabling block sources. Right now, this separation (keeping state in poller to figure out when to stop polling a block source) feels very tied to the idea that we'd only ever decide to disable a poller based on its response to a get_best_block() call. In the future, we may decide that a block source providing us garbage blocks is *also* ground for disabling it entirely, at which point this abstraction doesn't make sense anymore.

I see that putting all of the multi-source logic behind a single `BlockSource` multiplexer may take a bit more complexity, but I still think it wouldn't be as bad as you're thinking. It probably even makes sense to replace the recent headers cache with a simple limited-to-100k-entries HashMap or so - its only 100KB and covers a week of headers, then you don't even have to consider the complexity of clearing it when things are synced.

Reply:
The larger question is around where responsibility lies. I prefer to keep the responsibilities narrow as a means to separate concerns. So in this example, a `BlockSource` is only concerned with retrieving chain data. And implementations of `Poll` are only concerned with how to query block sources.

To support the use case that you mentioned, the `Poll` trait can be expanded with a `disable_source` function, which still falls under the concern of "how to query block sources". Then `MicroSPVClient` could signal to disable a block source if it provided garbage blocks. Such decisions may not be possible from within a `BlockSource` alone but rather may require the additional context provided by `MicroSPVClient` (e.g., disabling a source that is on a different chain).

Reply:
Thought about this some more after our offline discussion. I do see value in having a struct like `BlockSoureMultiplexer` in so far as it would allow users to customize policies around retrying, using backup sources, and generally choosing a source as it sees fit. However, such a multiplexer could accomplish this by implementing _both_ the `BlockSource` and `Poll` traits. Returning itself from `Poll::poll_chain_tip` is effectively how it would multiplex `get_block`. It could even decided to return a single source from `poll_chain_tip` in cases were it decides that is best.

As mentioned above, there is still a need to signal back somehow to say some source returned bad data. This is because there are two checks that `MicroSPVClient` performs that either require its context or it should be relied upon performing:
- [found a different chain](https://github.com/rust-bitcoin/rust-lightning/pull/763/files#diff-f457bab978fc8b89ad308d5195f99d7b65a4a6ba1673c5f164104b2dda9a0db6R183) during fork detection
- [sanity checking block data](https://github.com/rust-bitcoin/rust-lightning/pull/763/files#diff-f457bab978fc8b89ad308d5195f99d7b65a4a6ba1673c5f164104b2dda9a0db6R320) received from a source

The former can only be determined in the context of fork detection. The latter presumably should be handled by `MicroSPVClient` rather than relying on a `BlockSource` to implement this.

Agreed that the caching logic can be simplified which would clean up the `Poll` logic a bit. Backup querying would be removed since the multiplexer would handle it. Otherwise, I'm not sure if there is much else that can be removed from `MicroSPVClient`. Let me know what you think and if I missed something.

Reply:
As for your two checks, can't the former be detected in some other way? Basically just have the block source know about the chain's genesis or even pass to the block source in calls from the spv client. The second is obviously more nuanced - I'd generally think the block source could handle it, but you end up easily double-checking the data there, which isn't ideal. One design could be to make some container which holds the transaction list and does the checking there - the neat part there is it may line up with returning a list of transactions anyway, by allowing us to have a constructor that takes a full block (and does the checks) or one that takes a pre-filtered list of transactions.

Reply:
Alright, I managed to massage this into a cleaner design now, IMHO.

`Poll`'s interface has been expanded such that it is now a pure adapter of `BlockSource`, meaning neither it nor `MicroSPVClient` expose `BlockSource` as part of their interface any longer. Instead, functionality needed by `MicroSPVClient` is defined by `Poll`.

`ChainPoller` implements `Poll` for a single `BlockSource`, while `Poll`'s interface is defined in terms of `ValidatedBlockHeader` and `ValidatedBlock`. These essentially encapsulate the checks that `MicroSPVClient` had been performing previously. Therefore, all checks now occur directly within `ChainPoller`.

More elaborate implementations of `Poll` can and must be defined in terms of `ChainPoller`. For example, `ChainMultiplexer` is defined in this manner and no longer needs to implement `BlockSource`. 

Regarding the nuances around transaction lists, I think this can be part of a later refactor after we get a sample node working with a full-block source. Something like what you imagine should work well with the current design with minimal refactoring.

Thoughts?

Reply:
Looks good to me.